src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -909,51 +676,8 @@ private void WritePrimitiveType(Type type, EncodingKind kind)
 
         private void WriteType(Type type)
         {
-            int id;
-            if (_referenceMap.TryGetReferenceId(type, out id))
-            {
-                Debug.Assert(id >= 0);
-                if (id <= byte.MaxValue)
-                {
-                    _writer.Write((byte)EncodingKind.TypeRef_1Byte);
-                    _writer.Write((byte)id);
-                }
-                else if (id <= ushort.MaxValue)
-                {
-                    _writer.Write((byte)EncodingKind.TypeRef_2Bytes);
-                    _writer.Write((ushort)id);
-                }
-                else
-                {
-                    _writer.Write((byte)EncodingKind.TypeRef_4Bytes);
-                    _writer.Write(id);
-                }
-            }
-            else
-            {
-                _referenceMap.Add(type);
-
-                _writer.Write((byte)EncodingKind.Type);
-
-                TypeKey key;
-                if (!_binder.TryGetTypeKey(type, out key))
-                {
-                    throw NoSerializationTypeException(type.FullName);
-                }
-
-                this.WriteStringValue(key.AssemblyName);
-                this.WriteStringValue(key.TypeName);
-            }
-        }
-
-        private int _recursionDepth;
-        private const int MaxRecursionDepth = 50;
-
-        internal class RecursionDepthExceeded : Exception
-        {
-            public RecursionDepthExceeded()
-            {
-            }
+            _writer.Write((byte)EncodingKind.Type);
+            this.WriteInt32(ObjectBinder.GetTypeId(type));
```

Comment:
>ObjectBinder.GetTypeId(type) [](start = 28, length = 28)

Not all types written will have type id's.  Only the types that are serializable will have id's.   You'll need to have two different kinds of type encoding.